lib/smf/Kconfig

```diff
@@ -19,4 +19,12 @@ config SMF_INITIAL_TRANSITION
 	help
 	   If y, then each state can have an initial transition to a sub-state
 
+config SMF_UML_HSM
+	bool "Use UML Hierarchical State Machine rules"
+	select SMF_INITIAL_TRANSITION
+	select SMF_ANCESTOR_SUPPORT
+	help
+	   If y, then each transitions from parent state follow UML rules.
+	   smf_ctx objects use slightly more memory
```

Comment:
Won't this new Kconfig option cause issues with incompatibilities between state machines written for "classic" and others written for "UML" behaviour in the same code base?

Some subsystems within the tree use SMF and expect one behaviour. If the user enabled this Kconfig option to get a different behaviour for their application-specific state machines, this will effect _all_ SMF users in the tree as well.

Reply:
That's a very good point.

I could put in a function to enable UML behavior on an individual state machine context, with the default being Classic. something like `smf_set_uml_transitions(struct smf_ctx *ctx)`

This would mean every `smf_ctx` would have an `executing` member when `CONFIG_SMF_UML_HSM` is enabled, regardless of if it is UML or Classic, but if you're adding another HSM that uses UML transitions, you're most likely not in a resource-constrained situation and can afford the extra pointer.

Reply:
@glenn-andrews, greetings. From @henrikbrixandersen's perspective, wouldn't it be better to add e.g. a `smf_flags_t flags` parameter, where `typedef uint32_t smf_flags_t;`, to pass capabilities or behaviors to the target SMF via `smf_set_initial` and to set it in `smf_ctx`?
By doing this, we can maintain the same API while offering the user a choice of behavior during runtime.

The new `smf_set_initial` will look like this.

---
void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state, **smf_flags_t flags**);

Reply:
If I make a function you only call if you are wanting your state machine to enter UML mode, the signature of all classic functions, including `smf_set_initial()` will be unchanged. I'd rather not break other people's code.